runner/copy.go

```diff
@@ -75,28 +75,44 @@ func (m *MaterializedChunkRunner) Run() (types.CompletionWatcher, error) {
 			jobWatcher.EndWatch(fmt.Errorf("failed to create iterator: %w", err))
 			return
 		}
-		i := 0
+		// Buffer the records channel to ensure readers won't block on the writer loop below
+		ch := make(chan provider.ResourceRecord, 100_000)
+		var wg sync.WaitGroup
+		wg.Add(100)
+		// Create a "pool" of 100 goroutines to write to the table to better utilize CPU resources
+		// and speed up the write operations to the inference table.
+		for idx := 0; idx < 100; idx++ {
+			go func() {
+				defer wg.Done()
+				for record := range ch {
+					err := m.Table.Set(record.Entity, record.Value)
+					if err != nil {
+						close(ch)
+						jobWatcher.EndWatch(fmt.Errorf("could not set table: %w", err))
+						return
+					}
+				}
+			}()
+		}
+		// Completely consume the iterator and write all records to the channel
 		for it.Next() {
```

Comment:
i'm no goroutines expert, maybe this is standard, but it seems counterintuitive to start consuming from the channels first and produce to the channels below

Reply:
The logic is here:
* create a channel that has a buffer of 1M records, which means that sending isn't blocked by receiving
* spawn 1K go routines first so that once you start sending messages into the channel, receive and process them, theoretically keeping the actual buffer usage < 1M records
* Consume the iterator last because it's much faster than the Set operations performed in the go routines.

Reply:
I'm not sure I'm following and I'm asking because I'm trying to understand. This is what the code looks like to me

`
for all goroutines:
    read record from channels

for all records in iterator:
   put record in channel
`

it just feels more intuitive to do it the other way around? Is there some sort of performance benefit we get from doing it this way?

Reply:
If you send messages to the channel before you had go routines (i.e. workers to pull from it) you would synchronously fill the buffer, and exceed it in the case of files with 1M+ rows, before anything was initialized to process them.

The current pattern allows those go routines (workers) to begin working through the buffer concurrently as it's being filled by the iterator.

Reply:
ok simba just explained it to me

Reply:
Sorry, I just saw your explanation, that's exactly it!